### ISSUE TITLE
cluster-wise WB for deactivation seems to systematically yield empty results 

### ISSUE THREAD

**Issue by BryanGuillaume** [NISOx-BDI/SwE-toolbox#116]
While doing some testing when I was implementing the post-hoc masking for the WB, I have noticed that the cluster-wise WB for deactivation (i.e. negative effect) seems to systematically yield empty results. As proof that something is off, I have run twice the same WB analysis but have used the contrast [-1 0 1] for the first analysis and [1 0 -1] for the second analysis. Here below, the obtained -log10(WB clusterwise p-value) images:

### contrast [-1 0 1] (left: positive effect, right: negative effect)
<img width="1428" alt="-1 0 1" src="https://user-images.githubusercontent.com/4426961/58624944-acc6db80-82d1-11e9-854d-9a26f39004f4.png">

### contrast [1 0 -1] (left: positive effect, right: negative effect)
<img width="1429" alt="1 0 -1" src="https://user-images.githubusercontent.com/4426961/58624942-acc6db80-82d1-11e9-8c4f-e934ffd9feb2.png">

We can indeed see that the negative effect images are empty for the 2 analyses while they should show something similar than in the other analysis.

Also, looking at the saved maximum cluster sizes, we observe that the registered max bootstraped cluster sizes are always 0 for the negative effect, which is unlikely to be correct.

### max cluster sizes in SwE.mat for [-1 0 1]
maxClusterSize: [353 5 5 4 144 8 9 8 4 8 11 2 14 17 6 17 42 12 7 5]
maxClusterSizeNeg: [1717 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]

### max cluster sizes in SwE.mat for [1 0 -1]
maxClusterSize: [1717 4 13 29 12 6 3 33 6 11 10 22 0 2 2 77 2 41 2 48]
maxClusterSizeNeg: [353 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]

Also, this issue seems to be present in the test data because the negative effect image is also empty.

The issue is likely to be somewhere in swe_cp_wb. I will try to fix this today.

**Comment by BryanGuillaume** [comment]
I found the bug. For the bootstraps and image inputs, the variable indicating the voxels that were surviving the "negative" thresholding was initialised, but never updated:

### Initialisation of activatedVoxelsNeg
```
  if (SwE.WB.clusterWise == 1)
    activatedVoxels = false(1,S);
    if (SwE.WB.stat == 'T')
      activatedVoxelsNeg = false(1,S);
    end
  end
```

### Update of activatedVoxelsNeg missing here
```
      % hypothesis test
      if (WB.clusterWise == 1)
        hyptest=swe_hyptest(SwE, score, blksz, cCovBc, Cov_vis, dofMat);
        activatedVoxels(index) = hyptest.positive.activatedVoxels;
        clear cCovBc
      end
```
Thus, that explains why we got 0 as maximum bootstraped cluster sizes. Though, writting this, I just realise that this behaviour should have actually yielded results with the lowest p-value in each cluster... which was not the case as the result images seemed empty. Thus, there might be another bug somewhere in the code. Looking into it....

**Comment by BryanGuillaume** [comment]
Checking for a long time in the code, I could not see anything wrong in it that would explain why the result looked empty in FSLeye and in SPM. And I can actually confirm that the lowest p-values were actually saved into the result image. The issue seems to be that both FSLeye and SPM do not want to render these voxels at all. That can be seen in the screenshots below:

### Old results displayed in SPM (note that it shows that there is a value at the selected voxel - see Intensity field below the voxel coordinates- but it does not render it)
<img width="486" alt="SPM results" src="https://user-images.githubusercontent.com/4426961/58635582-882d2c80-82ee-11e9-8e61-da49c90ccf7a.png">

### Old (left) and new (right) results displayed in FSLeyes (note that it shows that there is a value at the selected voxel  for both, but render it only on the right image)
<img width="1431" alt="FSLeyes" src="https://user-images.githubusercontent.com/4426961/58635581-87949600-82ee-11e9-9db3-82bace3605a5.png">

The only special characteristic of these images is that they contains only nan or a single float value (here, 1.3...). I guess that in this scenario, FSLeyes and SPM cannot render this single float value as they cannot contrast it with a different value. One solution would be to change nan values by 0, but that could actually hide any cluster with p-value = 1, which may happen relatively easily. @nicholst, do you have an opinion about this issue?

**Comment by nicholst** [comment]
FSLeyes doesn’t play well with NaN’s. My recollection is that SPM uses NaN for masking but inconsistently - beta images getting NaN masking but statistic images getting 0 masking. 

If it makes our lives easier I’m happy to drop all NaN from output... only need to ensure that the mask.nii image is accurate and has the final, actual analysis mask (remembering that, due to all-constant voxels, some voxels that would be thought to be ‘in mask’ by the explicit mask and/or intensity masking aren’t in the actual analysis).

**Comment by BryanGuillaume** [comment]
The bug in the computation of clusterwise WB p-values for negative t-contrast has been resolved in  PR #117. As I am still not sure what to do regarding the "nan vs 0" issue and because it does not seem of high priority, I will open another issue regarding this and close this one
